std/digest/xxh.d

```diff
@@ -0,0 +1,3352 @@
+        }
+        while (input < limit);
+
+        h64 = XXH_rotl64(v1, 1) + XXH_rotl64(v2, 7) + XXH_rotl64(v3, 12) + XXH_rotl64(v4, 18);
+        h64 = XXH64_mergeRound(h64, v1);
+        h64 = XXH64_mergeRound(h64, v2);
+        h64 = XXH64_mergeRound(h64, v3);
+        h64 = XXH64_mergeRound(h64, v4);
+
+    }
+    else
+    {
+        h64 = seed + XXH_PRIME64_5;
+    }
+
+    h64 += cast(xxh_u64) len;
+
+    return XXH64_finalize(h64, input, len, align_);
+}
+
+XXH64_hash_t XXH64(const void* input, size_t len, XXH64_hash_t seed) @trusted pure nothrow @nogc
+{
+    static if (!XXH_NO_STREAM && XXH_SIZE_OPT >= 2)
+    {
+        /* Simple version, good for code maintenance, but unfortunately slow for small inputs */
+        XXH64_state_t state;
+        XXH64_reset(&state, seed);
+        XXH64_update(&state, cast(const(xxh_u8)*) input, len);
+        return XXH64_digest(&state);
+    }
+    else
+    {
+        if (XXH_FORCE_ALIGN_CHECK)
+        {
+            if (((cast(size_t) input) & 7) == 0)
+            { /* Input is aligned, let's leverage the speed advantage */
+                return XXH64_endian_align(cast(const(xxh_u8)*) input, len,
+                        seed, XXH_alignment.XXH_aligned);
+            }
+        }
+
+        return XXH64_endian_align(cast(const(xxh_u8)*) input, len, seed,
+                XXH_alignment.XXH_unaligned);
+    }
+}
+
+XXH64_state_t* XXH64_createState() @trusted pure nothrow @nogc
+{
+    import core.memory : pureMalloc;
+
+    return cast(XXH64_state_t*) pureMalloc(XXH64_state_t.sizeof);
+
+}
+
+XXH_errorcode XXH64_freeState(XXH64_state_t* statePtr) @trusted pure nothrow @nogc
+{
+    import core.memory : pureFree;
+
+    pureFree(statePtr);
+    return XXH_errorcode.XXH_OK;
+}
+
+void XXH64_copyState(XXH64_state_t* dstState, const XXH64_state_t* srcState) @trusted pure nothrow @nogc
+{
+    import core.stdc.string : memcpy;
+
+    memcpy(dstState, srcState, (*dstState).sizeof);
+}
+
+XXH_errorcode XXH64_reset(XXH64_state_t* statePtr, XXH64_hash_t seed) @trusted pure nothrow @nogc
+{
+    import core.stdc.string : memset;
+
+    assert(statePtr != null, "statePtr == null");
+    memset(statePtr, 0, (*statePtr).sizeof);
+    statePtr.v[0] = seed + XXH_PRIME64_1 + XXH_PRIME64_2;
+    statePtr.v[1] = seed + XXH_PRIME64_2;
+    statePtr.v[2] = seed + 0;
+    statePtr.v[3] = seed - XXH_PRIME64_1;
+    return XXH_errorcode.XXH_OK;
+}
+
+XXH_errorcode XXH64_update(XXH64_state_t* state, const void* input, size_t len) @trusted pure nothrow @nogc
+{
+    import core.stdc.string : memcpy;
+
+    if (input == null)
+    {
+        assert(len == 0, "len != 0");
+        return XXH_errorcode.XXH_OK;
+    }
+
+    {
+        const(xxh_u8)* p = cast(const(xxh_u8)*) input;
+        const xxh_u8* bEnd = p + len;
+
+        state.total_len += len;
+
+        if (state.memsize + len < 32)
+        { /* fill in tmp buffer */
+            memcpy((cast(xxh_u8*) state.mem64) + state.memsize, input, len);
```

Comment:
We should avoid using `memcpy` and friends and replace it with slice ops. Compiler generate bound checks to prevent errors on writing wrong `memcpy` calls.

Reply:
I just plain ported all the C code. Getting rid of <string.h> C legacy will be the next step :-)

Reply:
memcpy replaced with slice ops